### ISSUE TITLE
Private Route that redirects logged out User

### ISSUE THREAD

**Issue by jordantomax** [4Catalyzer/found#385]
Hello! 

**What's happening?**
I have an app that uses a custom `Route` class to redirect logged out users to `log-in`. It looks like this:

```
export class PrivateRoute extends Route {
  render ({ Component, props }) {
    if (!props) return undefined
    if (Component && props && !props.viewer) {
      throw new RedirectException('/log-in')
    }
    return (<Component {...props} />)
  }
}
```

This works most of the time, but if a user clicks a link twice, the props are defined but do not contain viewer, and so redirects.

**What I'd expect?**
Props contain the same data if a user clicks once or twice.

Do you have recommendations on how to handle this situation?

**Comment by taion** [comment]
That's weird! What's in `props`, if not `viewer`? Is a second query getting issued to your back end?

**Comment by jordantomax** [comment]
@taion Thanks for the quick reply! Yeah, it's weird!

There's only 1 query on the backend. Here's a screenshot from the network tab:

<img width="579" alt="Screenshot 2019-05-22 16 06 56" src="https://user-images.githubusercontent.com/37018/58205076-b9cd4480-7cab-11e9-94fb-aaf5d705934b.png">

Just 1 `OPTIONS` request and 1 `POST` request 🤔.

The props are packed full of other things that are not there on the 1st click. Here's screenshots of 1st and 2nd click. The 2nd one is the one with more props.

<img width="95" alt="Screenshot 2019-05-22 15 54 19" src="https://user-images.githubusercontent.com/37018/58205131-dd908a80-7cab-11e9-9324-b08f70e80e2d.png">
<img width="124" alt="Screenshot 2019-05-22 15 54 12" src="https://user-images.githubusercontent.com/37018/58205132-dd908a80-7cab-11e9-8409-3d936ba3de0f.png">

**Comment by jordantomax** [comment]
I should also mention (and you probably see) that I am using `found-relay` as well.

**Comment by jordantomax** [comment]
I'm curious under what circumstances the `routes` prop should be visible to this component. Adding another check for `&& !props.routes` does fix the problem, but it seems like a side effect.

**Comment by taion** [comment]
`props` should generally only be `match`, `router`, and the Relay props; see https://github.com/relay-tools/found-relay/blob/900a1555e2cac926b64bfada4a56215f3b7a66be/src/renderElement.js#L40

However, if the parent of this component is calling something like `cloneElement`, then it's possible for additional props to get injected down, and those will get forwarded through.

Is the `render` method in your `PrivateRoute` getting called as a _component_ "render" method, or is it getting called as a route's "render" method?

**Comment by jordantomax** [comment]
I'm not completely clear on the distinction, I think as a component `render`. It _was_ the child of another `Route`, but I replaced that with `Fragment` and nothing changed. It sits directly inside the outermost `Fragment` now. So basically it looks like:

```
export default makeRouteConfig(
  <Fragment>
    <LoginRoute exact path='/log-in' Component={Login} />

    <PrivateRoute path='/' Component={App} query={AppQuery}>
        // ...A bunch of routes
    </PrivateRoute>
  </Fragment>
)
```

My connected router setup is standard out of the box:

```
const ConnectedRouter = createConnectedRouter({
  render: createRender({
    renderError: ({ error }) => (
      <div>
        {error.status === 404 ? 'Not found' : 'Error'}
      </div>
    )
  })
})
```

Certainly no `cloneElement` going on in parent elements. I may try to take a look at the props using the source code of the library, but would definitely be interested in other ideas. Thanks for your help!

**Comment by taion** [comment]
Two things:

1. Where are you logging the "props" as above?
2. Can you take a look at the call stack? What's calling the `render` for the route?

**Comment by jordantomax** [comment]
I'm logging props inside the `PrivateRoute` component:

```
export class PrivateRoute extends Route {
  render ({ Component, props }) {
    if (!props) return undefined
    console.log(props)
    if (Component && props && !props.viewer && !props.routes) {
      throw new RedirectException('/log-in')
    }
    return (<Component {...props} />)
  }
}
```

Okay, so this is interesting. The call stack for a normal route looks like this:

<img width="340" alt="Screenshot 2019-05-23 12 33 52" src="https://user-images.githubusercontent.com/37018/58270272-5f89bd80-7d57-11e9-95d3-f7a56012787d.png">

It's worth noting that it's being called twice even with the correct props, does that mean it's double rendering?

When I put a conditional to check for `if (props.route)`, the call stack inside of there looks like this:

<img width="340" alt="Screenshot 2019-05-23 12 35 31" src="https://user-images.githubusercontent.com/37018/58270429-cdce8000-7d57-11e9-8a1b-154b34c48831.png">

Thank you again so much for trouble shooting with me!

**Comment by jordantomax** [comment]
`route.render` inside of `Resolvers.js` is getting called any time it goes through a custom `Route` component, which makes sense to me now. And it's smart enough not to re-render the `PrivateRoute` normally. But for some reason, clicking a link twice is causing that method to trigger again with those strange props.

**Comment by taion** [comment]
Can you check what `route` actually _is_ in this context, and verify that it's what you expect? From the call stack, it looks like it's hitting the "not a Relay route" branch.

**Comment by BorisTB** [comment]
tl;dr: compare with `props.viewer === null`, problem is in found-relay, not in found

Not sure if you ever solved this, but it's because of
this line: https://github.com/relay-tools/found-relay/blob/master/src/Resolver.js#L157
and this line: https://github.com/relay-tools/found-relay/blob/master/src/Resolver.js#L123
so it's a bug in found-relay

Resolver see that queries are the same as the last time, deletes them, than for some reason renders every matched component without query with `match` as the component's props

**Comment by taion** [comment]
@BorisTB I don't think that's right. We only read `this.lastQuerySubscriptions` inside the `updateQuerySubscriptions` method. And you'll note that we immediately rebind `this.lastQuerySubscriptions` in https://github.com/relay-tools/found-relay/blob/842d475ed4ed5d1c3596fc8f42bd5a591fa4e29e/src/Resolver.js#L142 anyway. The deletion is only for the cached previous version, and the value passed through is the newly calculated `querySubscriptions` anyway, per https://github.com/relay-tools/found-relay/blob/842d475ed4ed5d1c3596fc8f42bd5a591fa4e29e/src/Resolver.js#L45.

**Comment by BorisTB** [comment]
Maybe I’m missing something, but try to set log points in Resolver. It looks to me like this:
- 1st click
- component props and query subs are correct, components getting rendered with proper queries
- 2nd click
- queries are removed, the whole query subs array is filled with nulls
- components are rendered with props: match
- query subs array is refilled with proper queries
- components are rendered with proper values

I feel like those first two steps after 2nd click should not be there, but I’m not sure how to prevent it

**Comment by taion** [comment]
Which query subs array exactly? Again, `this.lastQuerySubscriptions` is only used within that method for caching. It doesn't get passed through to the actual routes.

**Comment by BorisTB** [comment]
Well, not really if you run it two times fast enough. Check this: https://jsfiddle.net/3sn82pd0/ (look at result in console)

I believe it will work if `this.lastQuerySubscriptions = querySubscriptions` is changed to `this.lastQuerySubscriptions = [...querySubscriptions]`

The thing is `this.lastQuerySubscriptions[i] = null;` inside the map function is pointing to the same `querySubscriptions` array that is currently being created.

**Comment by taion** [comment]
Fixed in #382
